cdap-api/src/main/java/co/cask/cdap/api/plugin/PluginClass.java

```diff
@@ -62,13 +64,21 @@ public PluginClass(String type, String name, String description, String classNam
     this.description = description;
     this.className = className;
     this.configFieldName = configfieldName;
-    this.properties = properties;
-    this.endpoints = endpoints;
+    this.properties = Collections.unmodifiableMap(new HashMap<>(properties));
+    this.endpoints = Collections.unmodifiableSet(new HashSet<>(endpoints));
+    this.requirements = requirements == null ? Collections.emptySet() :
+      Collections.unmodifiableSet(new HashSet<>(requirements));
+  }
+
+  public PluginClass(String type, String name, String description, String className, @Nullable String configfieldName,
```

Comment:
Where this constructor is being called? Why allow null?

Reply:
It seems like the deserializer will never call with `null`.

Reply:
Done.